File: go-sdk/pkg/observability/logging/pii_redactor.go
package logging

import (
	"regexp"
	"strings"
)

// ---------------------------------------------------------------------------
// Skip patterns — matches that fall inside these spans are NEVER treated as PII
// ---------------------------------------------------------------------------
//
// Rationale for each entry:
//
//  1. URLs          — numeric segments in S3 keys, CDN paths, query params, etc.
//  2. File paths    — Unix (/dir/12345_file.pdf) and Windows (C:\dir\12345.pdf)
//  3. UUIDs         — trace/request IDs that happen to contain long digit runs
//  4. Epoch / compact timestamps — 10-digit (Unix s), 13-digit (ms), 14-digit (YYYYMMDDHHmmss)
//  5. Semver / build strings — "0.0.1", "2.3.14-rc1"
//  6. Prefixed business IDs   — "ORD-123456789", "TXN987654321" (alpha prefix + digits)
//
// NOTE: Zip/postal codes that share the \d{3}-\d{2}-\d{4} SSN pattern cannot
// be distinguished from SSNs in free text.  Protection for those is handled at
// the key level via NonSensitiveFields ("zipcode", "zip", "pincode").

var skipPatterns = []*regexp.Regexp{
	// 1. http / https URLs
	regexp.MustCompile(`https?://[^\s"'<>]+`),

	// 2. Local and UNC file paths
	regexp.MustCompile(`(?i)(?:[a-z]:\\|\\\\|/)(?:[^\s"'<>]*)`),

	// 3. Full UUID  e.g. 6ba7b810-9dad-11d1-80b4-00c04fd430c8
	regexp.MustCompile(`\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b`),

	// 4. Epoch timestamps and compact datetimes
	//    10-digit Unix seconds | 13-digit ms | 14-digit YYYYMMDDHHmmss
	regexp.MustCompile(`\b(?:20\d{12}|19\d{11}|\d{13}|\d{10})\b`),

	// 5. Semantic version / build numbers  e.g. 1.0.0  2.3.14-rc1  v0.0.1
	regexp.MustCompile(`\bv?\d+\.\d+[\.\d\w-]*\b`),

	// 6. Prefixed business IDs  e.g. ORD-123456789012  TXN987654321098
	regexp.MustCompile(`\b[A-Za-z]{2,}-\d{6,}\b`),
}

// isSkippable returns true when the match at [start, end] inside input falls
// entirely within any skip-pattern span.
func isSkippable(input string, start, end int) bool {
	for _, sp := range skipPatterns {
		for _, span := range sp.FindAllStringIndex(input, -1) {
			if start >= span[0] && end <= span[1] {
				return true
			}
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// Non-sensitive field allowlist
// ---------------------------------------------------------------------------
//
// Values of keys in this map are structural / operational data — they are
// passed through RedactMap completely untouched (no pattern matching).
// This is the correct place to protect things like zip codes whose numeric
// format could collide with PII patterns.
//
// Add new keys here as your schema grows.

var NonSensitiveFields = map[string]bool{
	// Tracing / correlation
	"trace_id":       true,
	"span_id":        true,
	"request_id":     true,
	"correlation_id": true,
	"session_id":     true,

	// Timestamps
	"timestamp":  true,
	"created_at": true,
	"updated_at": true,
	"event_time": true,
	"deleted_at": true,

	// Application metadata
	"version":     true,
	"build":       true,
	"service":     true,
	"environment": true,
	"caller":      true,
	"level":       true,
	"message":     true,

	// Business / order identifiers (non-personal)
	"order_id":       true,
	"transaction_id": true,
	"invoice_id":     true,
	"ticket_id":      true,
	"product_id":     true,
	"item_id":        true,
	"merchant_id":    true,

	// Address structural fields
	// These also guard zip/postal codes from the SSN pattern.
	"zipcode": true,
	"zip":     true,
	"pincode": true,
	"city":    true,
	"state":   true,
	"country": true,
	"street":  true,

	// File / media references
	"template_url": true,
	"file_url":     true,
	"image_url":    true,
	"avatar_url":   true,
	"document_url": true,
	"url":          true,
	"uri":          true,
	"path":         true,
	"file_path":    true,

	// Generic safe fields
	"expiry":     true,
	"expires_at": true,
	"status":     true,
	"type":       true,
	"category":   true,
	"page":       true,
	"limit":      true,
	"offset":     true,
	"count":      true,
	"total":      true,
}

// IsNonSensitiveField checks if a field name is a known safe / structural field.
func IsNonSensitiveField(fieldName string) bool {
	return NonSensitiveFields[strings.ToLower(fieldName)]
}

// ---------------------------------------------------------------------------
// Core redaction engine
// ---------------------------------------------------------------------------

// redactSmart applies re+mask to every match in input that is NOT covered
// by any skip pattern.
func redactSmart(input string, re *regexp.Regexp, mask func(string) string) string {
	var buf strings.Builder
	lastIndex := 0
	for _, loc := range re.FindAllStringIndex(input, -1) {
		start, end := loc[0], loc[1]
		buf.WriteString(input[lastIndex:start])
		match := input[start:end]
		if isSkippable(input, start, end) {
			buf.WriteString(match) // not PII — preserve verbatim
		} else {
			buf.WriteString(mask(match))
		}
		lastIndex = end
	}
	buf.WriteString(input[lastIndex:])
	return buf.String()
}

// ---------------------------------------------------------------------------
// PIIRedactor
// ---------------------------------------------------------------------------

// PIIRedactor handles redaction of sensitive data in log messages.
type PIIRedactor struct {
	patterns []redactionPattern
}

type redactionPattern struct {
	re   *regexp.Regexp
	mask func(string) string
}

// NewPIIRedactor creates a new redactor with default FinTech patterns.
func NewPIIRedactor() *PIIRedactor {
	return &PIIRedactor{
		patterns: []redactionPattern{
			{
				// PAN Card (India): ABCDE1234F -> ABCDE****F
				re: regexp.MustCompile(`\b[A-Z]{5}[0-9]{4}[A-Z]\b`),
				mask: func(s string) string {
					if len(s) == 10 {
						return s[:5] + "****" + s[9:]
					}
					return s
				},
			},
			{
				// Aadhaar (India): 1234 5678 9012 -> XXXX XXXX 1234
				re: regexp.MustCompile(`\b(?:\d{4}\s?\d{4}\s?\d{4})\b`),
				mask: func(s string) string {
					rs := []rune(s)
					digitCount := 0
					for _, r := range rs {
						if r >= '0' && r <= '9' {
							digitCount++
						}
					}
					seen := 0
					for i, r := range rs {
						if r >= '0' && r <= '9' {
							seen++
							if seen <= digitCount-4 {
								rs[i] = 'X'
							}
						}
					}
					return string(rs)
				},
			},
			{
				// Credit/Debit Card: ************1234
				re: regexp.MustCompile(`\b(?:\d[ -]*?){13,19}\b`),
				mask: func(s string) string {
					rs := []rune(s)
					digitCount := 0
					for _, r := range rs {
						if r >= '0' && r <= '9' {
							digitCount++
						}
					}
					seen := 0
					for i, r := range rs {
						if r >= '0' && r <= '9' {
							seen++
							if seen <= digitCount-4 {
								rs[i] = '*'
							}
						}
					}
					return string(rs)
				},
			},
			{
				// Email: j***@domain.com
				re: regexp.MustCompile(`\b[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[A-Za-z]{2,}\b`),
				mask: func(s string) string {
					at := strings.Index(s, "@")
					if at > 0 {
						prefix := s[:at]
						if len(prefix) > 1 {
							return string(prefix[0]) + "***" + s[at:]
						}
						return "***" + s[at:]
					}
					return s
				},
			},
			{
				// Phone: ******3210
				re: regexp.MustCompile(`\b(?:\+?\d{1,3}[-\s]?)?(?:\(?\d{3,5}\)?[-\s]?)?\d{5}[-\s]?\d{5}\b`),
				mask: func(s string) string {
					rs := []rune(s)
					digitCount := 0
					for _, r := range rs {
						if r >= '0' && r <= '9' {
							digitCount++
						}
					}
					seen := 0
					for i, r := range rs {
						if r >= '0' && r <= '9' {
							seen++
							if seen <= digitCount-4 {
								rs[i] = '*'
							}
						}
					}
					return string(rs)
				},
			},
			{
				// Account number: ******7890
				re: regexp.MustCompile(`\b\d{9,18}\b`),
				mask: func(s string) string {
					if len(s) > 4 {
						return strings.Repeat("*", len(s)-4) + s[len(s)-4:]
					}
					return strings.Repeat("*", len(s))
				},
			},
			{
				// IFSC: SBIN0*****
				re: regexp.MustCompile(`\b[A-Z]{4}0[A-Z0-9]{6}\b`),
				mask: func(s string) string {
					if len(s) == 11 {
						return s[:5] + "******"
					}
					return s
				},
			},
			{
				// Passport: A******7
				re: regexp.MustCompile(`\b[A-Z][0-9]{7}\b`),
				mask: func(s string) string {
					if len(s) > 2 {
						return string(s[0]) + strings.Repeat("*", len(s)-2) + string(s[len(s)-1])
					}
					return s
				},
			},
			{
				// SSN: ***-**-6789
				re: regexp.MustCompile(`\b\d{3}-?\d{2}-?\d{4}\b`),
				mask: func(s string) string {
					rs := []rune(s)
					digitCount := 0
					for _, r := range rs {
						if r >= '0' && r <= '9' {
							digitCount++
						}
					}
					seen := 0
					for i, r := range rs {
						if r >= '0' && r <= '9' {
							seen++
							if seen <= digitCount-4 {
								rs[i] = '*'
							}
						}
					}
					return string(rs)
				},
			},
		},
	}
}

// Redact applies all PII patterns to input, skipping any match that falls
// inside a URL, file path, UUID, timestamp, version string, or prefixed ID.
func (r *PIIRedactor) Redact(input string) string {
	result := input
	for _, p := range r.patterns {
		result = redactSmart(result, p.re, p.mask)
	}
	return result
}

// RedactMap recursively redacts string values in a map.
//
//   - SensitiveFields    → always fully replaced with [REDACTED]
//   - NonSensitiveFields → passed through untouched (no pattern matching)
//   - Everything else    → pattern-matched via Redact()
func (r *PIIRedactor) RedactMap(data map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range data {
		switch {
		case IsSensitiveField(k):
			result[k] = "[REDACTED]"
		case IsNonSensitiveField(k):
			result[k] = v // structural / operational — pass through
		default:
			result[k] = r.RedactValue(v)
		}
	}
	return result
}

// RedactValue redacts a single value (handles strings, maps, slices).
func (r *PIIRedactor) RedactValue(v interface{}) interface{} {
	switch val := v.(type) {
	case string:
		return r.Redact(val)
	case map[string]interface{}:
		return r.RedactMap(val)
	case []interface{}:
		result := make([]interface{}, len(val))
		for i, item := range val {
			result[i] = r.RedactValue(item)
		}
		return result
	default:
		return v
	}
}

// ---------------------------------------------------------------------------
// Field classification helpers
// ---------------------------------------------------------------------------

// SensitiveFields are field names whose values must always be fully masked.
var SensitiveFields = map[string]bool{
	"password":      true,
	"secret":        true,
	"token":         true,
	"api_key":       true,
	"apikey":        true,
	"authorization": true,
	"cvv":           true,
	"pin":           true,
	"private_key":   true,
}

// IsSensitiveField checks if a field name is sensitive.
func IsSensitiveField(fieldName string) bool {
	return SensitiveFields[strings.ToLower(fieldName)]
}

File: python-sdk/observability/logging/pii_redactor.py
"""PII redaction for log messages."""

import re
from typing import Any, Dict, List, Set, Union

# ---------------------------------------------------------------------------
# Skip patterns — matches that fall inside these spans are NEVER treated as PII
# ---------------------------------------------------------------------------

SKIP_PATTERNS = [
    # 1. http / https URLs
    re.compile(r'https?://[^\s"\'<>]+'),
    # 2. Local and UNC file paths
    re.compile(r'(?i)(?:[a-z]:\\|\\\\|/)(?:[^\s"\'<>]*)'),
    # 3. Full UUID
    re.compile(r'\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b'),
    # 4. Epoch timestamps and compact datetimes
    re.compile(r'\b(?:20\d{12}|19\d{11}|\d{13}|\d{10})\b'),
    # 5. Semantic version / build numbers
    re.compile(r'\bv?\d+\.\d+[\.\d\w-]*\b'),
    # 6. Prefixed business IDs
    re.compile(r'\b[A-Za-z]{2,}-\d{6,}\b'),
]

def is_skippable(input_str: str, start: int, end: int) -> bool:
    """Check if the match at [start, end] falls entirely within any skip-pattern span."""
    for pattern in SKIP_PATTERNS:
        for match in pattern.finditer(input_str):
            if start >= match.start() and end <= match.end():
                return True
    return False

# ---------------------------------------------------------------------------
# Non-sensitive field allowlist
# ---------------------------------------------------------------------------

NON_SENSITIVE_FIELDS = {
    # Tracing / correlation
    "trace_id", "span_id", "request_id", "correlation_id", "session_id",
    # Timestamps
    "timestamp", "created_at", "updated_at", "event_time", "deleted_at",
    # Application metadata
    "version", "build", "service", "environment", "caller", "level", "message",
    # Business / order identifiers (non-personal)
    "order_id", "transaction_id", "invoice_id", "ticket_id", "product_id",
    "item_id", "merchant_id",
    # Address structural fields
    "zipcode", "zip", "pincode", "city", "state", "country", "street",
    # File / media references
    "template_url", "file_url", "image_url", "avatar_url", "document_url",
    "url", "uri", "path", "file_path",
    # Generic safe fields
    "expiry", "expires_at", "status", "type", "category", "page",
    "limit", "offset", "count", "total",
}

def is_non_sensitive_field(field_name: str) -> bool:
    return field_name.lower() in NON_SENSITIVE_FIELDS

# ---------------------------------------------------------------------------
# Sensitive field blocklist
# ---------------------------------------------------------------------------

SENSITIVE_FIELDS = {
    'password', 'secret', 'token', 'api_key', 'apikey',
    'authorization', 'cvv', 'pin', 'private_key',
}

def is_sensitive_field(field_name: str) -> bool:
    return field_name.lower() in SENSITIVE_FIELDS


# ---------------------------------------------------------------------------
# Core redaction engine
# ---------------------------------------------------------------------------

def redact_smart(input_str: str, pattern: re.Pattern, mask_func) -> str:
    """Applies re+mask to every match in input that is NOT covered by any skip pattern."""
    result = []
    last_index = 0
    
    # We find all matches first
    matches = list(pattern.finditer(input_str))
    
    for match in matches:
        start, end = match.span()
        
        # Append text before this match
        result.append(input_str[last_index:start])
        
        matched_text = match.group(0)
        
        if is_skippable(input_str, start, end):
            result.append(matched_text) # not PII — preserve verbatim
        else:
            result.append(mask_func(match)) # Apply mask function
            
        last_index = end
        
    result.append(input_str[last_index:])
    return "".join(result)


class PIIRedactor:
    """Redacts PII from log messages using regex patterns."""

    @staticmethod
    def _mask_pan(match: re.Match) -> str:
        s = match.group(0)
        if len(s) == 10:
             return s[:5] + "****" + s[9:]
        return s

    @staticmethod
    def _mask_aadhaar(match: re.Match) -> str:
        s = match.group(0)
        chars = list(s)
        digit_count = sum(1 for c in chars if c.isdigit())
        seen = 0
        for i, c in enumerate(chars):
            if c.isdigit():
                seen += 1
                if seen <= digit_count - 4:
                    chars[i] = 'X'
        return "".join(chars)

    @staticmethod
    def _mask_card(match: re.Match) -> str:
        s = match.group(0)
        chars = list(s)
        digit_count = sum(1 for c in chars if c.isdigit())
        seen = 0
        for i, c in enumerate(chars):
            if c.isdigit():
                seen += 1
                if seen <= digit_count - 4:
                    chars[i] = '*'
        return "".join(chars)

    @staticmethod
    def _mask_email(match: re.Match) -> str:
        s = match.group(0)
        at_index = s.find("@")
        if at_index > 0:
            prefix = s[:at_index]
            if len(prefix) > 1:
                return prefix[0] + "***" + s[at_index:]
            return "***" + s[at_index:]
        return s

    @staticmethod
    def _mask_phone(match: re.Match) -> str:
        s = match.group(0)
        chars = list(s)
        digit_count = sum(1 for c in chars if c.isdigit())
        seen = 0
        for i, c in enumerate(chars):
            if c.isdigit():
                seen += 1
                if seen <= digit_count - 4:
                    chars[i] = '*'
        return "".join(chars)

    @staticmethod
    def _mask_account(match: re.Match) -> str:
        s = match.group(0)
        if len(s) > 4:
            return "*" * (len(s) - 4) + s[-4:]
        return "*" * len(s)

    @staticmethod
    def _mask_ifsc(match: re.Match) -> str:
        s = match.group(0)
        if len(s) == 11:
            return s[:5] + "******"
        return s

    @staticmethod
    def _mask_passport(match: re.Match) -> str:
        s = match.group(0)
        if len(s) > 2:
            return s[0] + "*" * (len(s) - 2) + s[-1]
        return s

    @staticmethod
    def _mask_ssn(match: re.Match) -> str:
        s = match.group(0)
        chars = list(s)
        digit_count = sum(1 for c in chars if c.isdigit())
        seen = 0
        for i, c in enumerate(chars):
            if c.isdigit():
                seen += 1
                if seen <= digit_count - 4:
                    chars[i] = '*'
        return "".join(chars)

    def __init__(self):
        """Initialize the redactor with updated patterns and maskers."""
        self.patterns = [
            # PAN Card
            (re.compile(r'\b[A-Z]{5}[0-9]{4}[A-Z]\b'), self._mask_pan),
            # Aadhaar
            (re.compile(r'\b(?:\d{4}\s?\d{4}\s?\d{4})\b'), self._mask_aadhaar),
            # Credit/Debit Card
            (re.compile(r'\b(?:\d[ -]*?){13,19}\b'), self._mask_card),
            # Email
            (re.compile(r'\b[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[A-Za-z]{2,}\b'), self._mask_email),
            # Phone
            (re.compile(r'\b(?:\+?\d{1,3}[-\s]?)?(?:\(?\d{3,5}\)?[-\s]?)?\d{5}[-\s]?\d{5}\b'), self._mask_phone),
            # Bank Account
            (re.compile(r'\b\d{9,18}\b'), self._mask_account),
            # IFSC Code
            (re.compile(r'\b[A-Z]{4}0[A-Z0-9]{6}\b'), self._mask_ifsc),
            # Passport
            (re.compile(r'\b[A-Z][0-9]{7}\b'), self._mask_passport),
            # SSN
            (re.compile(r'\b\d{3}-?\d{2}-?\d{4}\b'), self._mask_ssn),
        ]

    def redact(self, input_str: str) -> str:
        """Apply all PII redaction patterns to a string, skipping safe zones."""
        result = input_str
        for pattern, mask_func in self.patterns:
            result = redact_smart(result, pattern, mask_func)
        return result

    def redact_dict(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Recursively redact PII from a dictionary."""
        result = {}
        for key, value in data.items():
            if is_sensitive_field(key):
                 result[key] = '[REDACTED]'
            elif is_non_sensitive_field(key):
                 result[key] = value # structural / operational — pass through
            else:
                 result[key] = self.redact_value(value)
        return result
    
    def redact_value(self, value: Any) -> Any:
        """Redact a single value based on type."""
        if isinstance(value, str):
            return self.redact(value)
        elif isinstance(value, dict):
            return self.redact_dict(value)
        elif isinstance(value, list):
            return [self.redact_value(item) for item in value]
        elif isinstance(value, tuple):
            return tuple(self.redact_value(item) for item in value)
        elif isinstance(value, set):
            return {self.redact_value(item) for item in value}
        return value


# Global redactor instance
_redactor = PIIRedactor()


def redact_log_event(event_dict: Dict[str, Any]) -> Dict[str, Any]:
    """Redact PII from a structlog event dictionary."""
    return _redactor.redact_dict(event_dict)


def get_redactor() -> PIIRedactor:
    """Get the global PII redactor instance."""
    return _redactor
